src/query/functions/aggregation/count_values.go

```diff
@@ -0,0 +1,229 @@
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package aggregation
+
+import (
+	"fmt"
+	"math"
+	"strconv"
+
+	"github.com/m3db/m3/src/query/block"
+	"github.com/m3db/m3/src/query/executor/transform"
+	"github.com/m3db/m3/src/query/functions/utils"
+	"github.com/m3db/m3/src/query/models"
+	"github.com/m3db/m3/src/query/parser"
+)
+
+const (
+	// CountValuesType counts the number of non nan elements with the same value
+	CountValuesType = "count_values"
+)
+
+// NewCountValuesOp creates a new take operation
+func NewCountValuesOp(
+	opType string,
+	params NodeParams,
+) (parser.Params, error) {
+	if opType != CountValuesType {
+		return baseOp{}, fmt.Errorf("operator not supported: %s", opType)
+	}
+
+	return newCountValuesOp(params, opType), nil
+}
+
+// countValuesOp stores required properties for take ops
+type countValuesOp struct {
+	params NodeParams
+	opType string
+}
+
+// OpType for the operator
+func (o countValuesOp) OpType() string {
+	return o.opType
+}
+
+// String representation
+func (o countValuesOp) String() string {
+	return fmt.Sprintf("type: %s", o.OpType())
+}
+
+// Node creates an execution node
+func (o countValuesOp) Node(
+	controller *transform.Controller,
+	_ transform.Options,
+) transform.OpNode {
+	return &countValuesNode{
+		op:         o,
+		controller: controller,
+	}
+}
+
+func newCountValuesOp(params NodeParams, opType string) countValuesOp {
+	return countValuesOp{
+		params: params,
+		opType: opType,
+	}
+}
+
+type countValuesNode struct {
+	op         countValuesOp
+	controller *transform.Controller
+}
+
+type bucketColumn []float64
+
+type bucketBlock struct {
+	columns      []bucketColumn
+	indexMapping map[float64]int
+}
+
+// Process the block
+func (n *countValuesNode) Process(ID parser.NodeID, b block.Block) error {
+	stepIter, err := b.StepIter()
+	if err != nil {
+		return err
+	}
+
+	params := n.op.params
+	meta := stepIter.Meta()
+	seriesMetas := utils.FlattenMetadata(meta, stepIter.SeriesMeta())
+	buckets, metas := utils.GroupSeries(
+		params.MatchingTags,
+		params.Without,
+		n.op.opType,
+		seriesMetas,
+	)
+
+	stepCount := stepIter.StepCount()
+	tempBlock := make([]bucketBlock, len(buckets))
+	for i := range tempBlock {
+		tempBlock[i].columns = make([]bucketColumn, stepCount)
+		tempBlock[i].indexMapping = make(map[float64]int, 10)
+	}
+
+	for columnIndex := 0; stepIter.Next(); columnIndex++ {
+		step, err := stepIter.Current()
+		if err != nil {
+			return err
+		}
+
+		values := step.Values()
+		for bucketIndex, bucket := range buckets {
```

Comment:
maybe consider moving out processing of a single bucket into a separate function to make it easier to understand

Reply:
Good call, looks better now